superduperdb/ext/anthropic/model.py

```diff
@@ -0,0 +1,101 @@
+import dataclasses as dc
+import typing as t
+
+import anthropic
+from anthropic import APIConnectionError, APIError, APIStatusError, APITimeoutError
+
+from superduperdb.container.component import Component
+from superduperdb.container.encoder import Encoder
+from superduperdb.container.model import PredictMixin
+from superduperdb.ext.utils import format_prompt, get_key
+from superduperdb.misc.retry import Retry
+
+retry = Retry(
+    exception_types=(APIConnectionError, APIError, APIStatusError, APITimeoutError)
+)
+
+KEY_NAME = 'ANTHROPIC_API_KEY'
+
+
+@dc.dataclass
+class Anthropic(Component, PredictMixin):
+    """Anthropic predictor."""
+
+    #: The model to use, e.g. ``'claude-2'``.
+    model: str
+
+    #: The identifier to use, e.g. ``'my-model'``.
+    identifier: str = ''
+
+    #: The version to use, e.g. ``0`` (leave empty)
+    version: t.Optional[int] = None
+
+    #: Whether the model takes context into account.
+    takes_context: bool = False
+
+    #: The encoder identifier.
+    encoder: t.Union[Encoder, str, None] = None
+
+    #: A unique name for the class
+    type_id: t.ClassVar[str] = 'model'
+
+    #: Keyword arguments to pass to the client
+    client_kwargs: t.Dict[str, t.Any] = dc.field(default_factory=dict)
+
+    def __post_init__(self):
+        self.identifier = self.identifier or self.model
+
+    @property
+    def child_components(self):
+        if self.encoder is not None:
+            return [('encoder', 'encoder')]
+        return []
+
+
+@dc.dataclass
+class AnthropicCompletions(Anthropic):
+    """Cohere completions (chat) predictor."""
+
+    #: Whether the model takes context into account.
+    takes_context: bool = True
+
+    #: The prompt to use to seed the response.
+    prompt: str = ''
+
+    @retry
+    def _predict_one(self, X, context: t.Optional[t.List[str]] = None, **kwargs):
+        if context is not None:
```

Comment:
Don't we do almost exactly the same elsewhere? Yeah, it might be too hard to extract a common function, just a note.

Also just noting: this pattern of writing exactly the same code twice, once `async`, is everywhere in Python, and IMHO they should have figured out a way to do that better, like `await_if_async` or something like that. :-)

One more note: the Google C++ codebase had an absolute ban on optional parameters in C++ and strongly discouraged them in Python!

The argument was simple: "these are two different functions".  Unfortunately, with everything already multiplied by four (async vs not, one vs not), this isn't so possible...